lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -271,6 +271,288 @@ static FailureOr<Literal> parseIntegerLiteral(MLIRContext *context,
   return lit;
 }
 
+/// Find the lowest-common-ancestor `lcaModule`, between `srcTarget` and
+/// `wireTarget`, and set `pathFromSrcToWire` with the path between them through
+/// the `lcaModule`.
+static LogicalResult
+findLCAandSetPath(AnnoPathValue &srcTarget, AnnoPathValue &wireTarget,
+                  SmallVector<InstanceOp> &pathFromSrcToWire,
+                  FModuleOp &lcaModule, ApplyState &state) {
+
+  auto srcModule = cast<FModuleOp>(srcTarget.ref.getModule());
+  auto wireModule = cast<FModuleOp>(wireTarget.ref.getModule());
+  // Get the path from top to `srcModule` and `wireModule`, and then compare the
+  // path to find the lca. Then use that to get the path from `srcModule` to the
+  // `wireModule`.
+  auto *top = state.instancePathCache.instanceGraph.getTopLevelNode();
+  if (srcModule == wireModule) {
+    lcaModule = srcModule;
+    return success();
+  }
+  lcaModule = cast<FModuleOp>(top->getModule());
+  auto wireInstancePathsFromTop =
+      state.instancePathCache.getAbsolutePaths(wireTarget.ref.getModule());
+  if (wireInstancePathsFromTop.size() > 1)
+    return wireTarget.ref.getOp()->emitError(
+        "cannot handle multiple paths to DataTaps wire");
```

Comment:
This check will eventually be too restrictive. The wiring transform approach to handling this is to determine an "owner" of each sink to exactly one source. However, I really don't like this approach as it can lead to ambiguity. An alternative approach approach is to instead entirely trust the `srctTarget` and `wireTarget` and compute the LCA using the information they have. This should work fine for now as data taps are always going to use the `toAbsoluteTarget` Chisel API. Given that, doing an LCA computation that uses information from the `instancePathCache` should be unnecessary.

Note that I would like to change Chisel (soon, but not blocking this PR) to add a `toRelativeTarget` API that would remove the need to ever do an LCA computation because a relative target can be trusted to have its root as the LCA. With that change, the `srcTarget` and `wireTarget` can be blindly used for ref type boring.

WDYT?

Reply:
I have updated the logic to use the `instances` if it exists, and use the absolute path, if it does not. We can enforce that the annotations must specify the absolute path, if there are multiple instances of a companion module. Otherwise, `getInstancePahts` can get the path.